FEATURE: Optimizer creates repairing policies in the TOSCA accepted by the Deployer

perezp opened the pull request:
This PR makes the Optimizer module create repairing policies in the syntax given in Section 7.3 of this document https://docs.google.com/document/d/1Z9RdaWn54IomgI-jCiw-tdutsCdDoHfk67PNUASOFmY/edit# .  
There is also a bit of refactoring of previous repeated code. 

In order to have a finished way to create these policies, for the next PRs it is still required to specify which modules can scale. Following the last directives, I will create the documentation of the module description, usage and run (I do not add it now due to lack of time as somebody may want to see the code for discussion in the next meeting).

kiuby88 commented on planner/optimizer/optimizer-core/src/main/java/eu/seaclouds/platform/planner/optimizer/Optimizer.java:
```diff
@@ -28,25 +28,28 @@
    public Optimizer() {
-      NUMBER_OF_PLANS_GENERATED = 5;
-      searchName = SearchMethodName.BLINDSEARCH;
+      this(5,SearchMethodName.BLINDSEARCH,0.5);
```
It could be better using Constants, for example `DAFAULT_NUMBER_OF_PLANS_GENERATED = 5`

perezp replied:
You're right. Changing it..

kiuby88 commented on planner/optimizer/optimizer-core/src/main/java/eu/seaclouds/platform/planner/optimizer/Optimizer.java:
```diff
@@ -62,13 +65,13 @@ public Optimizer(int num, SearchMethodName name) {
             previousPlans = outputPlans;
 
          } catch (Exception exc) {
             log.info(
```
better, log.warn?

kiuby88 commented on planner/optimizer/optimizer-core/src/main/java/eu/seaclouds/platform/planner/optimizer/heuristics/AbstractHeuristic.java:
```diff
@@ -310,17 +315,143 @@ public void checkQualityAttachedToSolutions(Solution[] solutions) {
    }
 
+   // TODO: The following code is repeated from HillClimb. A refactor will be
+   // necessary.
```
is it TODO valid still?

perezp replied:
Great! It's not valid anymore. This PR includes the refactor... 
Thank you

kiuby88 commented on planner/optimizer/optimizer-core/src/main/java/eu/seaclouds/platform/planner/optimizer/heuristics/Anneal.java:
```diff
@@ -48,7 +46,7 @@ public Anneal() {
 
       cloudOffers.sortDescendingPerformance();
-      Solution[] bestSols = super.findInitialRandomSolutions(cloudOffers, numPlansToGenerate);
+      Solution[] bestSols = super.findInitialRandomSolutions(cloudOffers, numPlansToGenerate, topology);
```
`super` is not necessary here

kiuby88 commented on planner/optimizer/optimizer-core/src/main/java/eu/seaclouds/platform/planner/optimizer/heuristics/BlindSearch.java:
```diff
@@ -50,7 +50,7 @@ public BlindSearch(int maxIter) {
       // This way may help for replanning, when even the first attempt for
       // solution will be based on the current deployment
-      Solution[] bestSols = findSolutions(null, cloudOffers,numPlansToGenerate);
+      Solution[] bestSols = super.findInitialRandomSolutions(cloudOffers,numPlansToGenerate,topology);
```
same above. You can use directly super methods avoiding `super`, right?

perezp replied:
Done. Thanks. 
As you said, it was not necessary. I had kept them only to avoid confusion while I was doing the refactoring of these methods.

kiuby88 commented:
Thanks @perezp.
Could you add some test to check the policies generation?

perezp commented:
I'm writing tests for the generated policies. Thanks @kiuby88

perezp commented:
@kiuby88 
I addressed your comments and I solved a bug which made travis fail. 
There are now a lot of commits, do you want me to squash now or after you check the comments and changes?

kiuby88 commented on planner/optimizer/optimizer-core/src/main/java/eu/seaclouds/platform/planner/optimizer/util/YAMLmodulesOptimizerParser.java:
```diff
@@ -350,4 +350,17 @@ private static QualityInformation parseQoS(Map<String, Object> qosInformation) {
+       Map<String,Object> moduleProperties= (Map<String,Object>) moduleInfo.get(TOSCAkeywords.MODULE_PROPERTIES_TAG); 
+       boolean moduleCanScale = (boolean) moduleProperties.get(TOSCAkeywords.MODULE_AUTOSCALE_PROPERTY); 
+       return moduleCanScale;
+      }catch(Exception e){
```
Avoiding Empty catch. For example, you could use log

kiuby88 commented on planner/optimizer/optimizer-core/src/test/java/eu/seaclouds/platform/planner/optimizerTest/OptimizerTOSCASeptember2015Test.java:
```diff
@@ -32,201 +32,228 @@
    }
 
-}
+   @Test(enabled = true)
```
`enable = true` is default value, so you could remove it. Really, I think you could move `@Test` tag to class leve.

kiuby88 commented:
+1

perezp commented:
Thanks for the review @kiuby88 . Merging....
